src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs

```diff
@@ -183,20 +197,97 @@ await using (pendingConnection.ConfigureAwait(false))
+            // Handshake stopped by QuicListener.DisposeAsync:
+            // 1. Dispose the connection and by that shut it down --> application error code doesn't matter here as this is a transport error.
+            // 2. Connection won't be handed out since listener has stopped --> do not propagate anything.
+
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Info(connection, $"{connection} Connection handshake stopped by listener");
+            }
+
+            await connection.DisposeAsync().ConfigureAwait(false);
+        }
+        catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested)
+        {
+            // Handshake cancelled by QuicDefaults.HandshakeTimeout, probably stalled:
+            // 1. Connection must be killed so dispose it and by that shut it down --> application error code doesn't matter here as this is a transport error.
+            // 2. Connection won't be handed out since it's useless --> propagate appropriate exception, listener will pass it to the caller.
+
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(connection, $"{connection} Connection handshake timed out: {oce}");
+            }
+
+            Exception ex = ExceptionDispatchInfo.SetCurrentStackTrace(new QuicException(QuicError.ConnectionTimeout, null, SR.Format(SR.net_quic_handshake_timeout, QuicDefaults.HandshakeTimeout), oce));
+            await connection.DisposeAsync().ConfigureAwait(false);
+            if (!_acceptQueue.Writer.TryWrite(ex))
+            {
+                // Channel has been closed, connection is already disposed, do nothing.
+            }
+        }
+        catch (Exception ex)
+        {
+            // Handshake failed:
+            // 1. Dispose the connection and by that shut it down --> application error code doesn't matter here as this is a transport error.
+            // 2. Connection cannot be handed out since it's useless --> propagate the exception as-is, listener will pass it to the caller.
+
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(connection, $"{connection} Connection handshake failed: {ex}");
+            }
+
+            await connection.DisposeAsync().ConfigureAwait(false);
+            if (!_acceptQueue.Writer.TryWrite(ex))
+            {
+                // Channel has been closed, connection is already disposed, do nothing.
+            }
+        }
+    }
+
     private unsafe int HandleEventNewConnection(ref NEW_CONNECTION_DATA data)
     {
         // Check if there's capacity to have another connection waiting to be accepted.
-        PendingConnection pendingConnection = new PendingConnection();
-        if (!_acceptQueue.Writer.TryWrite(pendingConnection))
+        if (Interlocked.Decrement(ref _pendingConnectionsCapacity) < 0)
         {
+            Interlocked.Increment(ref _pendingConnectionsCapacity);
```

Comment:
log? May be handy for debug. and may be worth of counter.

Reply:
Good point. For counters, we should probably do overall walk through S.N.Quic and design and expose more of them.

Reply:
Added logging.